specs/IFramePermissionRequested.md

```diff
@@ -0,0 +1,434 @@
+            return L"Camera";
+        case COREWEBVIEW2_PERMISSION_KIND_GEOLOCATION:
+            return L"Geolocation";
+        case COREWEBVIEW2_PERMISSION_KIND_NOTIFICATIONS:
+            return L"Notifications";
+        case COREWEBVIEW2_PERMISSION_KIND_OTHER_SENSORS:
+            return L"Generic Sensors";
+        case COREWEBVIEW2_PERMISSION_KIND_CLIPBOARD_READ:
+            return L"Clipboard Read";
+        default:
+            return L"Unknown resources";
+    }
+}
+
+static void PutHandled(ICoreWebView2PermissionRequestedEventArgs* args)
+{
+    // In the case of an iframe requesting permission, the default behavior is
+    // to first raise the PermissionRequested event off of the CoreWebView2Frame
+    // and invoke it's handlers, and then raise the event off the CoreWebView2
+    // and invoke it's handlers. However, If we set Handled to true on the
+    // CoreWebView2Frame event handler, then we will not raise the
+    // PermissionRequested event off the CoreWebView2.
+    wil::com_ptr<ICoreWebView2PermissionRequestedEventArgs2> args2;
+    CHECK_FAILURE(args->QueryInterface(IID_PPV_ARGS(&args2)));
+    if (args2)
+    {
+        CHECK_FAILURE(args2->put_Handled(true));
+    }
+}
+```
+
+## C#: Registering IFrame Permission Requested Handler
+```c#
+private WebView2 m_webview;
+Dictionary<Tuple<string, CoreWebView2PermissionKind, bool>, bool> m_cachedPermissions;
+
+// Example Use Case - an app may want to specify that all iframes should
+// create a message box to prompt the user for approval on a permission request.
+// The approval state could be cached so that future requests are automatically
+// handled if they have been requested previously.
+void RegisterIFramePermissionRequestedHandler()
+{
+    m_webview.CoreWebView2.FrameCreated += (sender, frameCreatedArgs) =>
+    {
+        // Checking for runtime support of CoreWebView2Frame.PermissionRequested
+        try
+        {
+            frameCreatedArgs.Frame.PermissionRequested += (frameSender, permissionArgs) =>
+            {
+                // Developer can obtain a deferral for the event so that the CoreWebView2
+                // doesn't examine the properties we set on the event args until
+                // after the deferral completes asynchronously.
+                CoreWebView2Deferral deferral = args.GetDeferral();
+
+                // We avoid potential reentrancy from running a message loop
+                // in the permission requested event handler by showing the
+                // dialog asynchronously.
+                System.Threading.SynchronizationContext.Current.Post((_) =>
+                {
+                    using (deferral)
+                    {
+                        var cachedKey = Tuple.Create(permissionArgs.Uri,
+                            permissionArgs.PermissionKind, permissionArgs.IsUserInitiated);
+
+                        if (m_cachedPermissions.ContainsKey(cachedKey))
```

Comment:
Can use TryGetValue to avoid double-lookup.

```csharp
if (m_cachedPermissions.TryGetValue(cachedKey, out value))
{
    permissionArgs.State = value ? CoreWebView2PermissionState.Allow : CoreWebView2PermissionState.Deny;
}
```

Reply:
Sounds good